### PR TITLE
Jakobus42/feat/server feedback

### PR DESCRIPTION
**Pull request by Jakobus42**
**Note: This PR is still work in progress.**

I just wanted to get your feedback on the current architecture of the api package.

the current state of the architecture is:
- **api.go:** creating a new api instance, running the api and shutting down the api.
- **api_test.go:** tests for all of the endpoints.
- **handlers.go:** handlers for all the the endpoints.
- **middleware.go:** currently only contains the authentication middleware.
- **router.go:** defines all routes and binds them to a specific handler
- **services.go:** currently only contains the grading service
---
### **Endpoints**
**all endpoints run under the router group /shortinette/v1 and require bearer authentication**

#### **Any**
- **`/modules/:id/:intra_login/grademe`**  
 Triggers grading process for a specific module identified by `:id` and `:intra_login`.

#### **POST**
- **`/webhook/grademe`**  
 Triggers grading process for a specific module identified the id and intraLogin extracted from the Github payload.

- **`/modules`**  
  Creates a new module.

- **`/participants`**  
  Creates a new participant.

#### **PUT**
- **`/modules`**  
  Updates an existing module.

- **`/participants`**  
  Updates an existing participant.

#### **GET**
- **`/modules`**  
  Retrieves all modules.

- **`/participants`**  
  Retrieves all participants.

- **`/modules/:id/:intra_login`**  
  Retrieves a specific module identified by `:id` and `:intra_login`.

- **`/participants/:intra_login`**  
  Retrieves a specific participant identified by `:intra_login`.

#### **DELETE**
- **`/modules/:id/:intra_login`**  
  Removes a specific module identified by `:id` and `:intra_login`.

- **`/participants/:intra_login`**  
  Removes a specific participant identified by `:intra_login`.
  
---
  
I would be happy for any suggestions on what to change and what is still needed:)

### CODE REVIEW

**Review by ifaoji — CHANGES_REQUESTED** [review]
Webserver occasionally returning 500. I'm not sure what's the exact cause yet, but sometimes requests which return a 200 return a 500 :thinking:

**Comment by winstonallo** [comment]
add: 
* bind mount for sqlite db in compose.yaml
* new endpoint for launching the short after adding the participants `/shortinette/v1/launch`
    * creates module template, and then all participant repos from it
* sha256 signature authentication middleware for github webhook - `/shortinette/webhook/grademe` does not belong to the same group as the other endpoints anymore, since it is not meant to be used directly by anything else than github
* template creation including all branches (reduces load when uploading traces)
* scheduler for unlocking new module every time `time.Now() % module.Duration == 0`

in summary, the app should be ready to finalize the project (minus some issues i assume we will find & fix on the way)

**Review by winstonallo — APPROVED** [review]
